avocado/utils/asset.py

```diff
@@ -199,127 +299,105 @@ def fetch(self):
-        if not self.asset_hash:
-            return True
-        if self._get_hash_from_file(asset_file) == self.asset_hash:
-            return True
-        else:
-            return False
+        :return: asset file if exists or None
+        :rtype: str or None
+        """
+        for cache_dir in self.cache_dirs:
+            cache_dir = os.path.expanduser(cache_dir)
 
-    def _get_local_file(self, url_obj, asset_file):
-        if os.path.isdir(url_obj.path):
-            path = os.path.join(url_obj.path, self.name)
-        else:
-            path = url_obj.path
+            for location in self.locations:
+                relative_dir = self._get_relative_dir(location)
+                asset_path = os.path.join(cache_dir,
+                                          relative_dir,
+                                          self.basename)
+
+                # To use a cached file, it must:
+                # - Exists.
+                # - Be valid (not expired).
+                # - Be verified (hash check).
+                if (os.path.isfile(asset_path) and
+                        not self._is_expired(asset_path, self.expire)):
+                    try:
+                        with FileLock(asset_path, 30):
+                            if self._verify(asset_path):
+                                return asset_path
+                    except Exception:  # pylint: disable=W0703
+                        exc_type, exc_value = sys.exc_info()[:2]
+                        LOG.error('%s: %s', exc_type.__name__, exc_value)
+        return None
 
-        with FileLock(asset_file, 1):
-            try:
-                os.symlink(path, asset_file)
-                self._compute_hash(asset_file)
-                return self._verify(asset_file)
-            except OSError as detail:
-                if detail.errno == errno.EEXIST:
-                    os.remove(asset_file)
-                    os.symlink(path, asset_file)
-                    self._compute_hash(asset_file)
-                    return self._verify(asset_file)
+    def get_metadata(self):
+        """
+        Returns metadata of the asset if it exists or None.
 
-    @staticmethod
-    def _is_expired(path, expire):
-        if expire is None:
-            return False
-        creation_time = os.lstat(path)[stat.ST_CTIME]
-        expire_time = creation_time + expire
-        if time.time() > expire_time:
-            return True
-        return False
+        :return: metadata
+        :rtype: dict or None
+        """
+        asset_path = self.find_asset_file()
+        if asset_path:
+            metadata_file_name = "%s_metadata.json" % self.basename
+            metadata_path = os.path.join(os.path.dirname(asset_path),
+                                         metadata_file_name)
+            if os.path.isfile(metadata_path):
+                with open(metadata_path, "r") as metadata_file:
+                    metadata = json.loads(metadata_file.read())
```

Comment:
Somewhere before I mentioned the use of `json.dump(fd, string)`. Likewise you can use 'json.load(fd)`